server/db.go

```diff
@@ -419,49 +643,198 @@ func (c *Server) rotateIssuers() (err error) {
+		issuer.IssuerType,
+		issuer.IssuerCohort,
+		issuer.MaxTokens,
+		issuer.Version,
+		issuer.ExpiresAt,
+		issuer.Buffer,
+		issuer.Duration,
+	)
+
+	// get the newly inserted issuer identifier
+	if err := row.Scan(&issuer.ID); err != nil {
+		tx.Rollback()
+		return fmt.Errorf("failed to get v3 issuer id: %w", err)
 	}
 
-	signingKeyTxt, err := signingKey.MarshalText()
-	if err != nil {
-		c.Logger.Error("Error marshalling signing key")
-		return err
+	if err := txPopulateIssuerKeys(c.Logger, tx, issuer); err != nil {
+		tx.Rollback()
+		return fmt.Errorf("failed to close rows on v3 issuer creation: %w", err)
 	}
+	queryTimer.ObserveDuration()
+	return tx.Commit()
+}
 
-	queryTimer := prometheus.NewTimer(createIssuerDBDuration)
-	rows, err := c.db.Query(
-		`INSERT INTO issuers(issuer_type, issuer_cohort, signing_key, max_tokens, expires_at, version) VALUES ($1, $2, $3, $4, $5, 2)`,
-		issuerType,
-		issuerCohort,
-		signingKeyTxt,
-		maxTokens,
-		expiresAt,
+// on the transaction, populate v3 issuer keys for the v3 issuer
+func txPopulateIssuerKeys(logger *logrus.Logger, tx *sqlx.Tx, issuer Issuer) error {
+	var (
+		duration *timeutils.ISODuration
+		err      error
 	)
+
+	if issuer.Version == 3 {
+		// get the duration from the issuer
+		duration, err = timeutils.ParseDuration(issuer.Duration)
+		if err != nil {
+			return fmt.Errorf("failed to parse issuer duration: %w", err)
+		}
+	}
+
+	// v1/v2 issuers only have a buffer of 1
+	if issuer.Version < 3 {
+		issuer.Buffer = 1
+	}
+
+	start := issuer.ValidFrom
+	i := 0
+	// time to create the keys associated with the issuer
+	if issuer.Keys == nil || len(issuer.Keys) == 0 {
+		issuer.Keys = []IssuerKeys{}
+	} else {
+		// if the issuer has keys already, start needs to be the last item in slice
+		start = issuer.Keys[len(issuer.Keys)-1].EndAt
+		i = len(issuer.Keys)
+	}
+
+	valueFmtStr := ""
+
+	var keys = []issuerKeys{}
+
+	// for i in buffer, create signing keys for each
+	for ; i < issuer.Buffer; i++ {
+		end := new(time.Time)
+		if duration != nil {
+			// start/end, increment every iteration
+			end, err = duration.From(*start)
+			if err != nil {
+				tx.Rollback()
+				return fmt.Errorf("unable to calculate end time: %w", err)
+			}
+		}
+
+		signingKey, err := crypto.RandomSigningKey()
+		if err != nil {
+			logger.Error("Error generating key")
+			tx.Rollback()
+			return err
+		}
+
+		signingKeyTxt, err := signingKey.MarshalText()
+		if err != nil {
+			logger.Error("Error marshalling signing key")
+			tx.Rollback()
+			return err
+		}
+
+		pubKeyTxt, err := signingKey.PublicKey().MarshalText()
+		if err != nil {
+			logger.Error("Error marshalling public key")
+			tx.Rollback()
+			return err
+		}
+
+		keys = append(keys, issuerKeys{
+			SigningKey: signingKeyTxt,
+			PublicKey:  string(pubKeyTxt),
+			Cohort:     issuer.IssuerCohort,
+			IssuerID:   issuer.ID,
+			StartAt:    start,
+			EndAt:      end,
+		})
+
+		if start != nil && !(*start).Equal(*issuer.ValidFrom) {
```

Comment:
I am not sure if these are possibly the wrong fields but I dont think this will ever evaluate to true due to the pointers assignments of start and issuer.ValidFrom at the beginning of the block and then end being assigned to start.

Reply:
this is saying if the transient loop variable "start" is not nil and the value of that transient start variable is  not equal to the value of the issuer.ValidFrom (start is only equal to valid from  on first iteration) then append a comma to the validfmtstr variable which is constructing the multi-value insert.

After the first iteration it will always start will never be equal to valid from again (as we assign it to end) and therefore after the first loop iteration we will always append a "," prior to the next set of values.

I believe this works as expected but will add a test to demonstrate.